### ISSUE TITLE
pkg_tar() does not include the runfile tree when taring.

### ISSUE THREAD

**Issue by chin33z** [bazelbuild/bazel#671]
When build with --dynamic_mode=fully, pkg_tar() only tars the "files" into the tarball. It does not include the shared library(located in the runfile tree) it needs. I would be nice if pkg_tar packing all the resource it needs. Thanks.

**Comment by kamalmarhubi** [comment]
This makes the docker_build rules a bit limited. Eg it won't work for either python_binary or sh_binary with dependencies.

**Comment by kamalmarhubi** [comment]
I was looking into this yesterday as I'd like `docker_build` to work with things that need runfiles as well, such as an `sh_binary` with `deps`.

The path handling makes this a bit unclear. Eg, if there is a `strip_prefix` (pkg_tar) or `data_path` (docker_build), a prefix of the path from repository root is removed. The runfiles tree would need to have the full path in order to be accessed by any depending binaries.

Should `strip_prefix` / `data_path` imply not including runfiles? Or should there be a symlink to listed `files` placed at `strip_prefix` / `data_path` with the actual files residing in a location with a complete runfiles tree?

Also, I wasn't sure how to do this operationally. Would `build_tar` create a runfiles manifest and call `build-runfiles` to build the tree?

**Comment by kamalmarhubi** [comment]
@damienmg ping since you seem to be the main author of these sets of rules.

**Comment by damienmg** [comment]
@mattmoor FYI (the original author of docker_build).

We have this problem inside Google too. It would be nice to be able to ship the runfiles tree. The runfiles tree live in a subtree %{target}.runfiles so it is not really affected by the strip_prefix / data_path argument (we should merge those btw).

I don't have a good opinion on wether we should dedup files by using symlinks or simply copy the whole runfiles tree.

Certainly deduping file looks nice, in which case I think we should decide on a base directory that would contains the files as is in the runfiles tree, maybe /execroot (we are considering having this path anyway in Bazel).

**Comment by kamalmarhubi** [comment]
> strip_prefix / data_path argument (we should merge those btw).

Forked as #684.

**Comment by kamalmarhubi** [comment]
> I don't have a good opinion on wether we should dedup files by using symlinks or simply copy the whole runfiles tree.

If people end up putting multiple binaries in a single Docker image / tar then symlinks would be a size win for something like Python.

> Certainly deduping file looks nice, in which case I think we should decide on a **base directory that would contains the files as is in the runfiles tree**

Not sure what you mean by the bolded part. You mean that the complete paths from the repository would be present?

> maybe /execroot (we are considering having this path anyway in Bazel).

Is there anything that the community can do on this or is it an ongoing design discussion? Also, would it be possibly to handle purely in Skylark?

**Comment by bsilver8192** [comment]
FWIW, I have custom Skylark rules for creating Debian packages (created before the Bazel ones were released) which include the runfiles. Just copying all of the files (without symlinks) is pretty easy and works fine for things like Python, but when you start including C++ shared libraries it wastes a lot of space. I've implemented sharing runfiles within each directory by putting them all in a directory named "runfiles" and then creating symlinks to that, but that gets pretty complicated and doesn't work in general.

I like the idea of somehow deduplicating the files with symlinks instead of just copying them, but I don't have any particularly good ideas on how to implement it.

**Comment by damienmg** [comment]
@kamalmarhubi I mean to have all the files of runfiles tree present in a path on the tar like /execroot (or maybe that could be configurable with an attribute) then create the symlink forest for the runfiles tree

The execroot prefix is an ingoing discussion in the Bazel team and we have an in-flight design doc that will be send out to bazel-discuss@ in the next weeks (probably the week before christmas).

**Comment by kamalmarhubi** [comment]
@bsilver8192 could you share
> I have custom Skylark rules for creating Debian packages (created before the Bazel ones were released) which include the runfiles.

Could you share the rules? I'd be happy to do something hackish for now and wait on the execroot thing @damienmg mentions to land for a real deduplicating solution.

**Comment by damienmg** [comment]
Sorry, when I said the execroot is an ongoing discussion is for having that as a prefix in the output_base (it will be mostly invisible to users, apart a new symlink bazel-execroot probably). And because we are considering the name execroot it was making sense for me to have the same.

However for deb packaging it might also make sense to remap it to some other location (/usr/share/execroot?)

**Comment by kamalmarhubi** [comment]
Would that result in conflicts if you install two deb packages that include the same runfiles?

**Comment by mikedanese** [comment]
Is there a simple temporary workaround for this? I imagine I could use a genrule to find the runfiles directory and tar it but I'm wondering if there is a better way.

**Comment by damienmg** [comment]
mikedanese: use a skylark rules to iterates over a runfiles tree to place their somewhere.

I believe it is a good intermediate solution to dupe all the files.

**Comment by wiktortomczak** [comment]
I was wondering if there's any update on including the runfiles tree? Is it possible to package a py_binary into a docker container? It's not clear to me how to use the docker_build rule for this purpose, per Kamal's first message. Thanks.

**Comment by yugui** [comment]
@damienmg 
> use a skylark rules to iterates over a runfiles tree to place their somewhere.

I tried to write such a skylark rule but runfiles were not available in the sandbox.
Do you have any suggestions?

**Comment by damienmg** [comment]
You need to add the runfiles to the list of inputs for the action.

**Comment by afking** [comment]
Is there an example generating of accessing runfiles?

**Comment by trainman419** [comment]
The lack of runfiles in pkg_tar has been a major pain point for my team, so I've put together an initial implementation that's about 90% complete: https://github.com/bazelbuild/bazel/pull/2687 . Comments welcome.

**Comment by ash2k** [comment]
`include_runfiles` is not documented here https://docs.bazel.build/versions/master/be/pkg.html#pkg_tar. Is it available already?

@codebreach if you use rules_docker then it is probably best to ask in their issue tracker. I found this issue which looks very much like the same problem https://github.com/bazelbuild/rules_docker/issues/284

**Comment by jin** [comment]
cc @aiuto (should this be moved to rules_pkg?)

**Comment by aiuto** [comment]
Sure, since we won't fix it here. 
There is a related issue there: https://github.com/bazelbuild/rules_pkg/issues/145
